src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,743 @@
+  501: 'Not Implemented', // RFC 7231 6.6.2
+  502: 'Bad Gateway', // RFC 7231 6.6.3
+  503: 'Service Unavailable', // RFC 7231 6.6.4
+  504: 'Gateway Timeout', // RFC 7231 6.6.5
+  505: 'HTTP Version Not Supported', // RFC 7231 6.6.6
+  506: 'Variant Also Negotiates', // RFC 2295 8.1
+  507: 'Insufficient Storage', // RFC 4918 11.5
+  508: 'Loop Detected', // RFC 5842 7.2
+  509: 'Bandwidth Limit Exceeded',
+  510: 'Not Extended', // RFC 2774 7
+  511: 'Network Authentication Required', // RFC 6585 6
+};
+
+export type DataWrittenEvent = {
+  index: number;
+  entry: WrittenDataBufferEntry;
+};
+
+// @ts-expect-error TS2720 net.Server inconsistencies.
+export class Server
+  extends EventEmitter
+  implements _Server, BaseWithHttpOptions
+{
+  [kConnectionsCheckingInterval]?: number;
+  [kUniqueHeaders]: Set<string> | null = null;
+
+  // Similar option to this. Too lazy to write my own docs.
+  // http://www.squid-cache.org/Doc/config/half_closed_clients/
+  // https://wiki.squid-cache.org/SquidFaq/InnerWorkings#What_is_a_half-closed_filedescriptor.3F
+  httpAllowHalfOpen = false;
+  timeout = 0;
+  maxHeadersCount: number | null = null;
+  maxRequestsPerSocket = 0;
+
+  requestTimeout: number = 0;
+  headersTimeout: number = 0;
+  requireHostHeader: boolean = false;
+  joinDuplicateHeaders: boolean = false;
+  rejectNonStandardBodyWrites: boolean = false;
+  keepAliveTimeout: number = 5_000;
+  port?: number;
+
+  constructor(options?: ServerOptions, requestListener?: RequestListener) {
+    super();
+
+    if (options != null) {
+      storeHTTPOptions.call(this, options);
+    }
+
+    if (typeof options === 'function') {
+      requestListener = options;
+      options = {};
+    } else if (options == null) {
+      options = {};
+    } else {
+      validateObject(options, 'options');
+    }
+
+    if (requestListener) {
+      this.on('request', requestListener);
+    }
+
+    this.on('listening', setupConnectionsTracking);
+
+    this[kUniqueHeaders] = parseUniqueHeadersOption(
+      options.uniqueHeaders as (string | string[])[]
+    );
+  }
+
+  close(): this {
+    httpServerPreClose(this);
+    if (this.port) {
+      portMapper.delete(this.port);
+    }
+    return this;
+  }
+
+  closeAllConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  closeIdleConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  setTimeout(
+    msecs?: number | ((socket: Socket) => void),
+    callback?: (socket: Socket) => void
+  ): this {
+    if (typeof msecs === 'function') {
+      callback = msecs;
+      msecs = undefined;
+    } else if (typeof msecs === 'number') {
+      this.timeout = msecs;
+    }
+
+    if (typeof callback === 'function') {
+      this.once('timeout', callback);
+    }
+    return this;
+  }
+
+  async #onRequest(
+    request: Request,
+    _env: unknown,
+    _ctx: unknown
```

Comment:
What's the plan on making the `_env` and `_ctx` available? Are we just going to depend on importable env and importable waitUntil? If so, that should be documented.